### PR TITLE
sql: enum-named connector options

### PR DESCRIPTION
**Pull request by sploiselle**
Proposes a change to the Kafka connection object to use keywords rather than free-form strings, as well as an extensible engine for extracting these configs.

If this general approach looks good, I'll create a set of tests tomorrow.

### Motivation

This PR adds a known-desirable feature. Part of #11504

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - `CREATE CONNECTION...FOR KAFKA BROKER` now uses a set of keywords to specify security parameters to connect to Kafka brokers via SSL.

### CODE REVIEW

File: src/sql-parser/src/ast/defs/ddl.rs [hunk 1]
```diff
@@ -545,11 +546,84 @@ impl<T: AstInfo> AstDisplay for DbzTxMetadataOption<T> {
...
+    SslKeyPassword,
+    SaslMechanisms,
+    SaslUsername,
+    SaslPassword,
+    SaslPasswordEnv,
+    GssapiKeytab,
+    GssapiKinitCmd,
+    GssapiReloginTime,
+    GssapiPrincipal,
+    GssapiService,
```
Comment (benesch) [inline src/sql-parser/src/ast/defs/ddl.rs:562]: No Kerberos in Cloud, so these can all evaporate!

```suggestion
```

File: src/sql-parser/src/ast/defs/ddl.rs [hunk 2]
```diff
@@ -545,11 +546,84 @@ impl<T: AstInfo> AstDisplay for DbzTxMetadataOption<T> {
...
 
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum KafkaConnectorOptionName {
+    SslKeyFile,
+    SslCertificateFile,
+    SslKeyPassword,
+    SaslMechanisms,
+    SaslUsername,
+    SaslPassword,
+    SaslPasswordEnv,
```
Comment (benesch) [inline src/sql-parser/src/ast/defs/ddl.rs:557]:
```suggestion
```

(No more env variants in Cloud!)

**Review by benesch — COMMENTED** [review]
Very neat! I don't quite know what option names we're going to want to use, but we basically want:

```sql
CREATE CONNECTION blah for KAFKA (
    SSL KEY db.schema.secret_ssl_key,
    SSL KEY PASSWORD db.schema.secret_ssl_key_password,
    SSL CERTIFICATE <string|secret>,
)
```

Then we'll use the `ssl.key.pem` librdkafka option rather than the `ssl.key.location` option to provide the key as a PEM-formatted string directly (reading it out of the secret value at the last second), rather than dealing with files.

**Comment by benesch** [comment]
```
CREATE CONNECTION blah for KAFKA (
    SSL KEY db.schema.secret_ssl_key,
    SSL KEY PASSWORD db.schema.secret_ssl_key_password,
    SSL CERTIFICATE <string|secret>,
)
```

It occurs to me this is likely to be a pain to parse because you have to disambiguate between `SSL KEY` and `SSL KEY PASSWORD`. Not sure how else to say `SSL KEY`, though. Hmm!

**Comment by sploiselle** [comment]
@benesch That is really not a problem!

```rust
match self.expect_one_of_keywords(&[KEY, CERTIFICATE])? {
            KEY => {
                if self.parse_keyword(PASSWORD) {
                    KafkaConnectorOptionName::SslKeyPassword
                } else {
                    KafkaConnectorOptionName::SslKey
                }
            }
````

**Comment by benesch** [comment]
Ok cool! I think the defect is that you can't use a secret named `password` without fully qualifying it or using an equals sign:

```sql
SSL KEY db.schema.password -- works
SSL KEY = password -- works
SSL KEY password -- oops, parsed as `SSL KEY PASSWORD` and then becomes a syntax error due to missing value 
```

But I'm totally willing to live with that!

**Comment by sploiselle** [comment]
@benesch This is just a v0 to get feedback. I can clean up `kafka_util` to just have one engine driving option extraction, but wanted to make sure this was on the right track before sinking a day into it.

File: src/sql-parser/src/parser.rs [hunk 3]
```diff
@@ -1925,7 +1925,7 @@ impl<'a> Parser<'a> {
             Keyword::Kafka => {
                 self.expect_keyword(BROKER)?;
                 let broker = self.parse_literal_string()?;
```
Comment (benesch) [inline src/sql-parser/src/parser.rs:1927]: I propose despecialifying the `BROKER` clause and making it just another option!

```sql
> CREATE CONNECTION kafka_ssl
  FOR KAFKA
      BROKER 'kafka:9092',
      SSL KEY = SECRET ssl_key_kafka,
      -- ...
```

It also occurs to me that `BROKERS = [...]` should be an array-accepting option that is mutually exclusive with `BROKER`. Then we could get rid of the gross accidental behavior that you can comma separate multiple addresses in the `BROKER` clause.

File: src/sql/src/kafka_util.rs [hunk 4]
```diff
@@ -40,6 +45,126 @@ enum ValType {
...
+        }
+    }
+
+    Ok(())
+}
+
+trait ConfigGen {
+    fn config_key(&self) -> &str;
+    fn config_val(&self) -> ValType;
+}
```
Comment (benesch) [inline src/sql/src/kafka_util.rs:98]: I'm not sure this generalizes beyond Kafka—none of the other connectors want their options in a bag of BTreeMap<String, StringOrSecret>, but instead want actual typed structs—so you might be able to just impl these directly on `KafkaConnectionOptionName`.

**Review by benesch — COMMENTED** [review]
Yeah, looks great! I've been thinking we probably want to change `KafkaConnection` to have explicitly typed fields for the few configuration parameters that we support, rather than this bag of `BTreeMap<String, StringOrSecret>` that could in principle contain any librdkafka option. That'll force us to be more disciplined about what options we accept. So I wouldn't invest _too_ much in making the Kafka configuration extraction general purpose. But I think this looks great wrt making incremental progress. 

If we mandate the use of connections, then we'll switch all the tests over to connections, and so no need to delete the old config parsing in `kafka_util`!

Btw, I have some changes coming down the pipe for the CSR connector. Going as fast as I can, but probably best not to start on the CSR syntax stuff until you see that PR!

**Comment by benesch** [comment]
#13013 is the PR, as promised. It's unfortunately not yet at the point where it compiles, but it's very close.

File: src/sql-parser/src/ast/defs/ddl.rs [hunk 5]
```diff
@@ -566,10 +625,27 @@ impl<T: AstInfo> AstDisplay for CreateConnection<T> {
             } => {
                 f.write_str("KAFKA BROKER '");
                 f.write_node(&display::escape_single_quote_string(broker));
-                f.write_str("'");
-                if with_options.len() > 0 {
-                    f.write_str(" WITH (");
-                    f.write_node(&display::comma_separated(&with_options));
+                f.write_str("' ");
+
+                // TODO(move this into a trait with a generic implementation)
```
[the remark this answers is not in the document]
Comment (sploiselle) [reply src/sql-parser/src/ast/defs/ddl.rs:630]: Ah, I hadn't realized that. I think the group is something that seems nice but makes a huge mess. I'm in favor of the syntax you listed and will move the implementation in that direction.

File: src/sql/src/kafka_util.rs (hunk 4 above)
Comment (sploiselle) [reply src/sql/src/kafka_util.rs:108]: This has to be a trait to impl them in this crate, and I had anticipated doing the same thing for the CSR Connector. I hear you about making these structs, though--if this is all right to get this one PR over the line, I'll adjust the subsequent PRs to move in that direction.

Comment (benesch) [reply src/sql/src/kafka_util.rs:108]: Oh d'oh, they're SQL parser structs. No yeah totally fine, sorry!

**Comment by sploiselle** [comment]
@benesch PTAL; I don't think this is ideal, but moves things forward a bit!

**Comment by benesch** [comment]
I'm in meetings for a while yet, but this is in my queue!

**Comment by sploiselle** [comment]
Supereded by #13161

**Pull request closed without merging**